[Spring Data JPA] 김진학 미션 제출합니다.

iampingu99 opened the pull request:
## 🏃‍♂️ 과정

JpaRepository에서 기본으로 제공하는 메서드와 네이밍 컨벤션을 사용하여 데이터베이스 조회를 하도록 했습니다. 기존 schema.sql 파일의 DDL과 `@Entity` 로 인한 테이블 생성이 중복되어 데이터 삽입을 위한 DML 만을 포함한 data.sql 파일로 분리하게 되었습니다.

이번 리뷰도 잘 부탁드립니다.

## **📃 요구사항**

### JPA 전환

**로직**

- [x]  Member / Reservation / Theme / Time JPA 전환 및 연관관계 매핑
- [x]  초기 데이터 쿼리 파일 수정

**예외**

- [x]  파라미터로 Theme / Time 를 찾을 수 없는 경우

## 🔎 궁금한점
    
1. 레포지토리, 서비스, 컨트롤러 로직 분리
    
    정답은 없겠지만 오답이나, 선호하시는 방법에 대한 의견 부탁드립니다.
    
    1. 주키로 엔티티를 조회하는 것은 빈번하며 예외를 한번에 처리하기 위해서 서비스에 `findByOrElse()` 등의 메서드를 두었습니다. 몇몇개는 한줄 코드를 메서드로 분리 (`save()` , `deleteById()`) 하고 있는데 과도해 보이기도 하네요.
    2. `서비스`에서 `컨트롤러`의 요청 객체 `HttpServletRequest` , `XXXRequest`  또는 반환 객체 `XXXResponse` 를 서비스가 알 필요가 없다고 생각합니다. 서비스는 레포지토리와 서비스간의 엔티티 또는 DTO 에만 의존하는 것은 어떨까요?
    3. 서비스 로직으로 분리하고 자주 사용하다 보니, 다른 서비스 객체에서 서비스 객체에 의존해서 사이클이 발생하기도 했습니다. (ex) MemberService 와 ReservationService 가 서로 의존하는 경우) 이 경우 잘못된 방식으로 분리하게 되어서 발생하는 문제인 것 같은데 그럼 서비스에서 서비스 로직을 의존하기보단, XXXRepository 메서드를 직접 사용하는 것이 더 좋은 것인가요?

iampingu99 committed:
chore(config): JPA 전환을 위한 의존성 추가 및 스키마 삭제

자동으로 테이블을 생성해주는 기능 사용으로 스키마 삭제

BREAKING CHANGE: schema.sql 파일은 제거됨

iampingu99 committed:
refactor(time): JPA 전환

iampingu99 committed:
refactor(theme): JPA 전환

iampingu99 committed:
refactor(reservation): JPA 전환

iampingu99 committed:
refactor(member): JPA 전환

iampingu99 committed:
feat(config): JPA 초기 데이터 data.sql 추가

IDENTITY 전략 사용

iampingu99 committed:
test(time): 4단계 JPA 전환 테스트 추가

iampingu99 committed:
feat(time): 조회 결과가 없는 경우 예외 추가

iampingu99 committed:
feat(theme): 서비스 레이어 분리 및 조회 결과가 없는 경우 예외 추가

youngsu5582 commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -1,26 +1,31 @@
         this.role = role;
     }
 
+    public Member() {
```
해당 생성자가 왜 필요한가요?

iampingu99 replied:
잘못 이해하고 있거나, 이상한 부분은 말씀해주세요 👀

### Spring Boot 3.1.0 Hibernate 가 enetity 를 생성하는 흐름

1. 엔티티를 인스턴스화 하기 위한 객체인 `EntityInstantiatorPojoStandard` 의 생성자 함수에서 엔티티의 생성자 함수를 가져온다.
   
- `EntityInstantiatorPojoStandard.EntityInstantiatorPojoStandard`
    <img width="859" alt="10" src="https://github.com/user-attachments/assets/97f938fc-9ce7-467e-913e-623c4fa532bd" />

- `EntityInstantiatorPojoStandard.resolveConstructor`
    <img width="852" alt="11" src="https://github.com/user-attachments/assets/87595a8f-cd88-45cb-935d-e41dac4391a3" />
    
   
    
2. `ReflectHelper.getDefaultConstructor` 에서 `Class.getDeclaredConstructor()` 로 기본 생성자를 가져온다. 추가로 `setAccessibility()` 로 기본 생성자가 `private` 접근 제어자를 가지더라도 접근할 수 있도록 해준다.

- `ReflectHelper.getDefaultConstructor`
    <img width="924" alt="13" src="https://github.com/user-attachments/assets/a698201d-b320-4846-bdcc-0f3c5122d178" />
    
- `ReflectHelper.ensureAccessibility`
    <img width="924" alt="14" src="https://github.com/user-attachments/assets/4727292d-6a80-4936-a0b6-e207563671d6" />
    

    
4. 기본 생성자를 사용하여 엔티티의 인스턴스를 생성한다.

-  `EntityInstantiatorPojoStandard.instantiate`
    <img width="924" alt="15" src="https://github.com/user-attachments/assets/cc5f5ee5-cb8a-4084-bacf-b73fb65a7c17" />
    

    

### 결론

엔티티의 인스턴스를 생성할때 Hibernate 는 Reflection 을 이용하여 `접근 제어자와 관계없이 엔티티의 기본생성자를 필요`로 한다.

youngsu5582 replied:
정리 되게 잘 해주셨네용

`ensureAccessibility` 을 통해 접근 권한을 가능하게 하면, 빈 생성자가 private 접근자여도 될 거 같은데 
JPA 는 왜 기본 생성자 접근 제한을  private 으로는 못 하게 할까요??

iampingu99 replied:
Entity 를 생성할 때에는 문제가 발생하지 않았습니다. 에디터에서 `클래스 'Member'에는 [public, protected] no-arg 생성자가 포함되어야 합니다.` 는 메세지를 출력하고 있지만, 위의 코드 흐름에 따라서 기본 생성자의 접근자를 private 로 변경해도 테스트를 통과하고 잘 동작했습니다. `하지만 Proxy 를 생성할 때 문제가 발생하게 되었습니다.`

### Spring Boot 3.1.0 Hibernate 가 proxy 를 생성하는 흐름
1. `ByteBuddyProxyFactory.getHibernateProxyInternal()`
    <img width="989" alt="image" src="https://github.com/user-attachments/assets/362a3a26-9dca-4da3-b833-9ce44a86491b" />
  프록시 객세를 생성하기 위해 ByteBuddyProxyFactory 가 생성자 함수를 통해 인스턴스를 생성하려고 합니다.

2. `Class.getConstructor()`

    <img width="806" alt="image" src="https://github.com/user-attachments/assets/1cd934f8-40b2-4420-914b-8904a338f786" />
    이후 Class 가 public 속성을 넣어 기본 생성자 함수 찾으려 합니다

3. Class.getDeclaredConstructors0
      <img width="860" alt="image" src="https://github.com/user-attachments/assets/d8ed1f8f-4eec-4d2b-9687-38dd45562c70" />

    앞선 public 속성 때문에 Member.class 를 통해서 프록시 객체를 생성할 public 기본생성자 만을 가져오는 듯 했으나, 마지막 함수 호출에서 `매개변수가 true 임에도 불구하고 public, protected 생성자 함수를 가져오게 되었습니다.` (그 이상의 코드는 확인할 수 없었습니다)

### 결론
표준에는 어긋나지만 Entity 를 생성하는 경우에 스프링 JPA 기본 구현체로 Hibernate 를 사용할때 private 접근자가 허용되는 것 같습니다. 하지만 Proxy 가 생성될 때는 private 접근자는 허용되지 않습니다. 즉, JPA 에서 기본 생성자 함수를 public 또는 protected 로 사용해야 합니다.

youngsu5582 replied:
그러면 스프링에서 프록시 객체를 생성하는 방식을 뭐라고 하나요?

iampingu99 replied:
탐구해본 결과 스프링에서 프록시 객체를 생성하는 방식으로 두가지가 있었습니다.
1. JDK 동적 프록시: 인터페이스 기반으로 프록시 객체를 생성하는 방식으로, 인터페이스를 구현한 클래스에서만 사용할 수 있습니다.
2. CGLIB: 바이트코드 조작을 통해 클래스를 상속하여 프록시 객체를 생성하는 방식으로, 인터페이스를 구현하지 않은 클래스에도 적용할 수 있습니다. 그러나 final 클래스는 상속이 불가능하므로 CGLIB을 사용할 수 없으며, private 메서드는 오버라이딩이 불가능하여 프록시의 대상이 될 수 없습니다.

하이버네이트의 프록시 객체 또한 위의 코드를 통해 인터페이스가 아닌 클래스 정보를 통해 객체를 생성하는 것으로 보입니다. 추가로 프록시 객체를 직접 구현하거나 AOP 를 사용해보지 못해서 이해하지 못한 부분들이 많습니다. 혹시 도움되는 자료나 지식을 알려주실 수 있을까요?

youngsu5582 replied:
https://www.youtube.com/watch?v=MFckVKrJLRQ
해당 테코톡을 보면 도움이 될 거 같네요.
추가로, AOP 는 깊게 사용하기 전에는 그냥 부가 기능을 붙여주는구나 로만 생각해도 될 거 같습니다.
( 왜, JPA 가 DB에 쿼리를 보내거나, 자동으로 JOIN 해오는지에 까지만 )

iampingu99 replied:
테코톡을 통해 프록시 객체 방식을 이해하는데 많은 도움이 되었습니다.
지금 직접 프록시나 AOP를 구현해보기 보다는, 말씀해주신대로 부가기능을 붙여준다는 생각으로 내부 흐름을 이해해봐야겠습니다. 감사합니다 👍

youngsu5582 commented on src/main/java/roomescape/auth/AuthService.java:
```diff
@@ -3,20 +3,21 @@
 public class AuthService {
     private final JwtTokenProvider jwtTokenProvider;
-    private final MemberDao memberDao;
+    private MemberRepository memberRepository;
```
해당 요소는 final 을 안 붙인 의도가 있나요?

iampingu99 replied:
아닙니다. final 을 빠트렸습니다. 수정하겠습니다 👍 

추가로 에디터가 필드가 final 이 될 수 있다고 추천해주기도 했고, 싱글톤을 위해 final 을 사용했습니다. 하지만 스프링 자체에서 싱글톤으로 관리해주는 것을 알게되었는데, 의도치 않은 재할당을 막아줄 수 있으므로 사용하는 것이 좋겠지요?

youngsu5582 replied:
넵. 사소한 습관이긴 한데 다른 사람들한테 의구심을 가지게 하는 컨벤션은 통일하는게 좋은거 같아용

youngsu5582 commented on build.gradle:
```diff
@@ -15,7 +15,8 @@ repositories {
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
-    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+//    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
```
사용하지 않으면 제거해도 될 거 같아요 👍

iampingu99 replied:
반영해보겠습니다 👍

youngsu5582 commented on src/main/java/roomescape/theme/ThemeService.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public Theme findByIdOrThrow(Long themeId) {
+        return themeRepository.findById(themeId)
```
성능상 큰 차이는 없을꺼 같은데
`String.format` 이 더 깔끔할 거 같네용

iampingu99 replied:
`String.format` 을 사용한다면 매개변수가 많아졌을 때도 직관적으로 보일 것 같네요 반영해보겠습니다 👍

youngsu5582 commented on src/main/java/roomescape/time/Time.java:
```diff
@@ -24,4 +35,8 @@ public Long getId() {
     public String getValue() {
         return value;
     }
+
```
getValue 가 있는데 getTime 을 만드는 이유가 있나요?

iampingu99 replied:
테스트로 주어진 코드 JpaTest.java 에서 `getTime()` 을 통해 검증하고 있었기에 새 메서드를 작성하게 되었습니다.

테스트에서 사용하기 위해 getter, setter 메서드는 지양해야 하는 것으로 알고 있습니다. 하지만 미션 테스트 코드 통과를 목적으로 두다보니 추가하게 되었습니다. 미션 테스트 코드를 수정해야 되겠지요?

youngsu5582 replied:
넵 기존 코드는 과감히 바꿔도 좋습니당

youngsu5582 commented on src/main/java/roomescape/time/TimeRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface TimeRepository extends CrudRepository<Time, Long> {
```
findAll 이 기본으로 제공 되지 않나요?

iampingu99 replied:
CrudRepository<T, ID> 는 Iterable<T> findAll 을 제공해줍니다. 하지만 Iterable 보다 List 가 더 익숙하기도 했고, 테스트시에 용이했기에 재정의해서 사용하게 되었습니다.

관련 자료를 찾아본 결과 `컨트롤러가 바디에 데이터를 Iterable<T> 타입로 반환하더라도 jackson 에 의해 직렬화 되어 List 와 같은 Json 형태가 되는 것`을 알게되었습니다. findAll 재정의로 인해 CrudRepository 내부적으로 Iterable 을 List 로 변경하는 비용이 생길 것 같은데 현재는 그대로 사용하는 것이 더 좋아보입니다.

추가로 List 를 사용하고 싶다면 CrudRository 이외에 ListCrudRepository 를 사용하는 방식을 고려해볼 수 있습니다.

youngsu5582 replied:
이것도 팀 내 컨벤션 및 저울질 일거 같네요.
`Iterable` 이라는게 불편해서 메소드를 만들거면 `ListCrudRepository` 를 안 쓰는 이유가 있긴 싶네용
( 어차피, `CrudRepository` 에서 `findAll()`,`findAllById(Iterable<ID> ids)` 만 추가되는게 다이기 때문 )

iampingu99 replied:
여러 종류의 Repository가 제공되는 만큼, 상황에 맞게 선택하여 활용하는 것이 중요해 보입니다. 나중에 팀 컨벤션을 정할 때, 적절한 방식을 제안할 수 있도록 잘 정리해두겠습니다 👍

youngsu5582 commented on src/main/java/roomescape/waiting/Waiting.java:
```diff
@@ -0,0 +1,62 @@
+    public String getDate() {
+        return date;
+    }
+
```
해당 부분에서 객체를 getter 로 가져오면 무슨 일이 일어나나요?

youngsu5582 commented on src/test/java/roomescape/JpaTest.java:
```diff
@@ -0,0 +1,31 @@
+import roomescape.time.Time;
+import roomescape.time.TimeRepository;
+
+@DataJpaTest
```
해당 어노테이션이 하는 역할은 뭔가요?

iampingu99 replied:
`@SpringBootTest` 와 다르게 전체 애플리케이션 컨텍스트를 로드하지 않고 JPA 계층의 컨텍스트를 로드하여 더 빠르게 수행할 수 있습니다. 또한 엔티티를 테스트하기 위한 `EntityManager`, `TestEntityManager` 객체를 제공해주며, 기본적으로 테스트가 끝난 후 데이터베이스 트랜잭션이 롤백되어 독립적으로 만들어줍니다.

youngsu5582 replied:
그러면 EntityManager 라는게 뭔가요?
왜 해당 로직에선 `entityManager.persist(...)` 와 `entityManager.flush()` 를 해줘야만 테스트가 성공하나요?

iampingu99 replied:
EntityManager는 데이터베이스와 엔티티 간의 상호작용을 관리합니다.
1. persist(entity)를 호출하면 엔티티가 영속화 즉, 영속성 컨텍스트에 등록되지만 즉시 데이터베이스에 반영되지는 않습니다. 따라서, persist() 이후 바로 데이터베이스에서 조회하는 테스트를 수행하면 실패할 수 있습니다.
2. flush()를 호출하면 영속성 컨텍스트의 변경사항이 데이터베이스에 즉시 반영됩니다. 이를 통해 데이터베이스에 저장된 엔티티를 정상적으로 조회할 수 있습니다. 추가로 엔티티를 조회할 때 해당 엔티티가 이미 영속화 되어있다면, 데이터베이스를 조회하는 대신, 영속성 컨텍스트에서 즉시 반환됩니다. 즉, 이미 영속성 컨텍스트에 존재하는 엔티티를 조회하면 추가적인 쿼리가 발생하지 않습니다.

youngsu5582 commented on src/main/java/roomescape/theme/Theme.java:
```diff
@@ -1,6 +1,15 @@
 public class Theme {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
IDENTITY 가 의미하는게 뭔가요?

iampingu99 replied:
`IDENTITY` 전략의 경우 키 생성 방식을 DB에 의존하는 것으로, DB의 스키마에서 지정한 방식과 같아지는 것입니다.
`AUTO` 전략의 경우 JPA가 DB 밴더에 알맞게 키 생성 방식을 정해주는 것입니다.

youngsu5582 replied:
그러면, H2 랑 MySQL 에서 기본 생성 방식은 뭔가요??

iampingu99 replied:
H2 의 기본 키 생성 방식은 SEQUENCE 이며 MySQL 은 AUTO_INCREMENT 입니다. 이때 `AUTO` 전략을 사용하게 되면 데이터베이스에 따른 기본 생성 방식을 따르게 됩니다.

youngsu5582 replied:
그러면 AUTO_INCREMENT 의 단점은 뭐가 있나요??

iampingu99 replied:
AUTO_INCREMENT 의 단점으로는 다음과 같습니다.
순차적으로 증가하는 키로 메모리도 적고 직관적이지만 보안 문제에 대한 단점이 있습니다. 사용자의 PK 나, 서비스의 데이터 크기 등을 예측할 수 있게 됩니다. 또한 삭제된 아이디를 재사용하지 않으므로 사용되지 않는 키들이 많아질 수 있습니다.

youngsu5582 commented on src/main/java/roomescape/time/TimeService.java:
```diff
@@ -31,14 +30,19 @@ public List<AvailableTime> getAvailableTime(String date, Long themeId) {
+        return timeRepository.findAll();
+    }
+
+    public Time findByIdOrThrow(Long id) {
```
interface 의 default 메소드로도 해결이 가능할거 같네요.
( 해당 메소드 그대로, interface 에 넣으면 됩니당 )

iampingu99 replied:
레이어 분리만 생각하고 default 메서드를 생각하지 못했네요. 반영해보겠습니다 👍

youngsu5582 reviewed (changes requested):
쓰다가 내용이 날라갔네요...
그래서, 내용이 조금 더 간단해집니다. 죄송합니당

> 주키로 엔티티를 조회하는 것은 빈번하며 예외를 한번에 처리하기 위해서 서비스에 `findByOrElse()` 등의 메서드를 두었습니다.

해당 내용은 리뷰에 남긴 것처럼 인터페이스에서 공통적으로 처리할 수 있습니다.
근데, 이를 호출하는 로직마다 다르게 처리해야 한다면 Optional 을 사용하는게 좋을거 같습니다.

> `서비스`에서 `컨트롤러`의 요청 객체 `HttpServletRequest` , `XXXRequest` 또는 반환 객체 `XXXResponse` 를 서비스가 알 필요가 없다고 생각합니다.

이를 위해서 Service 단 DTO 를 만드는 경우가 있습니다. ( 완벽한 격리 )
근데, 단순히 Service 단 DTO 를 만든다기 보단 서비스가 가져야 하는 객체를 만든다라고 생각하면 편할거 같습니다.
```kotlin
data class LocalLoginHttpRequest(
    val email: String,
    val password: String
)
```

```kotlin
data class MemberIdentifier(
    val email: String,
    val password: String
)
```
이와 같을때, 매개변수는 같지만 이름을 통해 서비스가 가져야 하는 객체임을 명확히 드러냅니다.
( 단순, Request Response 보다 더 직관적이기도 하구요 )

이렇게 하면, 웹 요청이 어떻게 변하든 변화되지 않는 서비스 코드가 됩니다.

> 서비스는 레포지토리와 서비스간의 엔티티 또는 DTO 에만 의존하는 것은 어떨까요?
해당 부분에서 엔티티는 다소 위험할 수 있습니다. `LazyinitializationException` 을 발생시킬 수 있기 때문입니다. ( 이에 대해 탐구하고 설명해주세용 )
추가로, 도메인 간 분리를 위해 ID 를 통해 조회를 하는 것도 일종의 비즈니스 로직 행위 중 하나일 수 있습니다. ( 기존, 다른 서비스를 통해 이미 알더라도 )

> 서비스 로직으로 분리하고 자주 사용하다 보니, 다른 서비스 객체에서 서비스 객체에 의존해서 사이클이 발생하기도 했습니다. 
이렇게 서비스가 서비스를 가지는걸 서비스 파사드 패턴이라 하는데요.
저는, 왠만하면 서비스가 다른 서비스를 가지는걸 피하는게 좋다고 생각합니다.
서비스는 정말 말그대로 비즈니스 로직을 수행하는 레이어인데
`멤버를 조회한다` 이게 비즈니스 로직을 수행한다고 생각하기 보단, 이를 수행하는 행위중 하나 라고 생각합니다.

그래서, `구현 계층(?)` 이라는 요소를 만들어 단순 로직 및 비즈니스 로직을 지탱하는 반복 행위 를 분리합니다.

[참고](https://geminikims.medium.com/%EC%A7%80%EC%86%8D-%EC%84%B1%EC%9E%A5-%EA%B0%80%EB%8A%A5%ED%95%9C-%EC%86%8C%ED%94%84%ED%8A%B8%EC%9B%A8%EC%96%B4%EB%A5%BC-%EB%A7%8C%EB%93%A4%EC%96%B4%EA%B0%80%EB%8A%94-%EB%B0%A9%EB%B2%95-97844c5dab63)

```kotlin
@Post("/api/auth/login")
fun login(@Body localLoginHttpRequest: LocalLoginHttpRequest): ApiResponse<LocalLoginHttpResponse> {
    val memberData = memberService.readMember(localLoginHttpRequest.toIdentifier())
    val token = tokenService.createToken(memberData.id)
    return apiResponse {
        data = LocalLoginHttpResponse(accessToken = token.accessToken.value)
    }
}
```
```kotlin
    fun readMember(memberIdentifier: MemberIdentifier): MemberData =
        MemberData.from(memberReader.findMember(memberIdentifier))
```
```kotlin
    fun findMember(memberIdentifier: MemberIdentifier): Member =
        (memberRepository.findByEmail(memberIdentifier.email)
            ?: throw NoSuchElementException("이메일(${memberIdentifier.email})에 해당하는 회원을 찾을 수 없습니다"))
            .also {
                if (passwordEncoder.matches(memberIdentifier.password, it.getPassword()).not()) {
                    throw IllegalArgumentException("비밀번호가 일치하지 않습니다")
                }
            }
```
샘플 코드입니다. ( 코틀린이나, 크게 어려움 없을거 같은데 추가 설명 필요하면 말해주세용 )

하지만, 이 역시도 오히려 간단한 코드나 확장성이 보이지 않을때 하면 더 복잡하거나 번거롭게 할 수 있습니다. 그렇기에 팀이나 자신만의 명확한 기준을 가져나가면 좋을거 같습니다. 👍

iampingu99 commented:
답변 감사합니다. 리뷰내용을 바탕으로 제가 이해한 것을 정리해보았습니다. 

1. Repository 의 조회 결과가 존재하지 않는 경우 Service 단에서 예외 반환 로직 중복

    ```
        memberRepository.findById(id).orElseThrow(...); // Member 를 조회 해야하는 로직이 필요한 모든 Service 레이어에서 중복 발생
    ```
     Repository interface 에 `default` 메서드를 구현하여 예외 반환 로직 중복을 제거할 수 있다. 또한 다른 예외 반환 로직이 필요한 경우 Optional 반환 메서드를 사용할 수 있다.

2. Repository 메서드를 Service 에 비즈니스 로직 없이 단순 위임

    ```
    public Member save(Member time) {
        return memberRepository.save(time);
    }
    ```
    확장성을 고려하거나, Controller 가 Service 에 의존하는 일관성을 유지하기 위해서 작성할 수 있다고 생각합니다. 이때 다른 Service 에서 해당 Service 메서드를 사용하는 것은 지양하거나, 구현 계층을 넣는 방식으로 사이클을 피할 수 있습니다.

3. Service 가 Entity 에 의존할때 발생할 수 있는 문제점
  `LazyInitializationException` 예외 발생: Hibernate가 Lazy Loading 된 프록시 객체의 실제 데이터를 조회하려할 때, 영속성 컨텍스트가 종료된 상태에서 시도하려는 경우 발생합니다. 예를 들면 영속성 컨텍스트가 종료된 후 컨트롤러에서 엔티티의 프록시 객체를 Json 으로 직렬화 하거나, 서비스에서 엔티티를 Dto 로 매핑 할 경우 프록시 객체의 실제 데이터 조회를 시도하고 해당 예외가 발생하게 됩니다. 
  
    Service 가 Entity 에 의존할 수 있으나 LazyInitializationException 예외에 대한 주의를 해야한다로 이해하면 될까요?
  
4. 해당 부분이 잘 이해가 가지 않습니다. 예시를 들어주실 수 있으실까요?
    > 추가로, 도메인 간 분리를 위해 ID 를 통해 조회를 하는 것도 일종의 비즈니스 로직 행위 중 하나일 수 있습니다. ( 기존, 다른 서비스를 통해 이미 알더라도 )

iampingu99 committed:
refactor(repository): 예외 중복 메서드 Repository default 메서드로 분리

time / theme findByIdOrElse 메서드 추가

iampingu99 committed:
refactor(reservation): service 의존을 repository 으로 변경

theme / time serivce 의존성을 theme / time repository 로 변경

BREAKING CHANGE: ThemeService.findByIdOrElse() 는 제거됨
TimeService.findByIdOrElse() 는 제거됨

iampingu99 committed:
refactor(reservation): 모든 예약 목록 조회 기능 N+1 문제 수정

lazy loading 및 fetch join 사용

youngsu5582 commented:
- 1번

정리하신게 맞습니다. 결론적으로, 해당 요소가 없는걸 DB Layer 가 처리할지, Service Layer 가 처리할지의 차이입니다.

- 3번

일반적으로 서비스가 메소드 전체에 트랜잭션을 붙이는 경우면 상관없으나,
외부 API 를 보내는 동안 트랜잭션을 의도적으로 끊는 경우도 있습니다. ( DB 커넥션 점유를 너무 오랫동안 점유하는 문제 발생 )

이런 경우, 아무 생각없이 트랜잭션이 끊어졌는데 이 객체를 서비스 매개변수로 담아서 보내면 문제가 발생합니다.

```kotlin
val lottoPublish = getPublishWithStatus(lottoPublishId,LottoPublishStatus.WAITING)
orderValidator.checkOrderValid(lottoPurchaseRequest.toOrderDataRequest())
val purchase = purchaseProcessor.purchase(lottoPurchaseRequest.toPurchaseRequest())
lottoPublisher.complete(lottoPublish)

@Transaction
@Write
fun complete(lottoPublish:LottoPublish): LottoPublish {
    lottoPublish.complete()
    return lottoPublish
}

fun complete() {
    if (this.status != LottoPublishStatus.WAITING) {
        throw IllegalStateException("결제 대기 상태에서만 완료가 가능합니다")
    }
    this.status = LottoPublishStatus.COMPLETE
}
```

이와 같을 시, lottoPublish 는 JPA 영속이 끊기고 ( Detached ) 
상태 변화를 하면 이게 DB 에 반영이 안됩니다.

이런점은 예외도 안던지므로, 더욱 심각한 문제가 될 수 있습니다.

- 4번

4번은 3번과도 연관되는 요소이기도 합니다. ( 객체로 보내면, 트랜잭션이 끊길수 있으므로 )
또는, `lotto` 라는 도메인과 `purchase` 라는 도메인이 있을때
로직을 통해 Lotto 와 Purcahse 라는 도메인을 둘다 알지만 패키지 및 도메인 영역 분리를 위해서

```kotlin
val purchase = purchaseProcessor.purchase(lottoPurchaseRequest.toPurchaseRequest())
val lottoPublish = lottoPublisher.complete(lottoPublishId)
return LottoBillData.from(lottoWriter.saveBill(purchase.getId(), lottoPublish.getId(), authenticated.memberId))
```

와 같은 접근도 가능할 수 있습니다. ( `LottoWriter 라는 객체는 굳이 Purchase 라는 도메인을 몰라도 된다` )

iampingu99 committed:
refactor(time): 중복되는 메서드 제거

getValue() 와 getTime() 의 동작은 같으므로 getTime() 제거

BREAKING CHANGE: getTime() 은 제거됨

iampingu99 reviewed (commented):
답변 감사합니다! 답변 내용을 바탕으로 이해한 내용을 정리해보았습니다.

```kotlin
val lottoPublish = getPublishWithStatus(lottoPublishId,LottoPublishStatus.WAITING) // lottoPublish 객체를 가져오고 트랜잭션이 끊어짐
orderValidator.checkOrderValid(lottoPurchaseRequest.toOrderDataRequest())
val purchase = purchaseProcessor.purchase(lottoPurchaseRequest.toPurchaseRequest())
lottoPublisher.complete(lottoPublish) // lottoPublish 는 Detached 상태

@Transaction // 트랜잭션이 생기지만 lottoPublish 는 영속성 컨텍스트에 존재하지 않는다.
@Write
fun complete(lottoPublish:LottoPublish): LottoPublish {
    lottoPublish.complete()
    return lottoPublish
}

fun complete() {
    if (this.status != LottoPublishStatus.WAITING) {
        throw IllegalStateException("결제 대기 상태에서만 완료가 가능합니다")
    }
    this.status = LottoPublishStatus.COMPLETE // 엔티티의 상태를 변경하지만 반영되지 않으며 이는 예외도 발생시키지 않으므로 문제가 될 수 있다
}
```

Service 가 Entity 를 반환하는 경우 트랜잭션이 끊기고 난 후 프록시 객체를 조회하려 하거나, 엔티티를 수정하는 것과 같이 트랜잭션이 필요한 동작을 시도한다면 문제가 발생할 수 있다. 

이를 방지하기 위해, 한 번에 연관 객체를 모두 조회하거나 DTO로 변환하여 반환하면 트랜잭션이 종료된 후에도 안전하게 데이터를 조회할 수 있다. 또한, 엔티티의 상태를 변경해야 하는 경우 트랜잭션을 유지하여 변경 사항이 정상적으로 반영되도록 해야 한다.

```kotlin
val purchase = purchaseProcessor.purchase(lottoPurchaseRequest.toPurchaseRequest())
val lottoPublish = lottoPublisher.complete(lottoPublishId)
return LottoBillData.from(lottoWriter.saveBill(purchase.getId(), lottoPublish.getId(), authenticated.memberId))
// LottoBillData.from(lottoWriter.saveBill(purchase, lottoPublish.getId(), authenticated.memberId))
```

이미 Purchase 객체를 알고 있으므로 Purchase 객체를 매개변수로 넘겨줄 수 있지만, LottoWriter 객체는 Purchase 객체를 알아야 한다. 이때, id 를 매개변수로 넘겨줌으로써 LottoWriter 에 id 를 통해 Purchase 를 찾도록 해서 도메인 영역 및 패키지를 분리할 수도 있다.

youngsu5582 commented on src/test/java/roomescape/reservation/ReservationServiceTest.java:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.context.annotation.Import;
+
+@DataJpaTest
+@Import(ReservationService.class)
```
DataJpaTest 가 의도가
```
JPA 계층의 컨텍스트를 로드하여 더 빠르게 수행할 수 있습니다.
```
라고 설명해주셨는데
서비스 계층의 Service 를 Import 해서 테스트를 하는 이유가 있나요?

iampingu99 replied:
mock 을 사용하기 보다는 데이터베이스와 연결된 테스트가 더 좋을 것이라 생각하여 통합테스트로 구성하였습니다. 하지만 `@SpringBootTest` 는 전체 컨텍스트를 로드하기 때문에 너무 과했기에 `@DataJpaTest` 를 통해서 JPA 계층의 컨텍스트만을 로드하려 했습니다. 

다음으로 서비스 로직을 그대로 옮겨와 테스트 결과를 확인해볼 수 있었지만 Service 테스트인 만큼 메서드를 사용하는 것이 좋다고 생각하여 필요한 Service 컨텍스트만 추가로 로드해서 사용하게 되었습니다.

youngsu5582 replied:
두 가지에서 선택을 해야할 거 같은데

제일 처음 얘기한 것처럼 결국 테스트에 트랜잭션 어노테이션을 붙이면, 기존 로직과 다르게 동작하는 결과를 초래할 수 있습니다.
대신, 이와같이 필요한 것만 로딩하면 더 빠르게 효율적으로 할 수 있겠죠.
( 추가로, 이는 스프링 테스트 컨텍스트 개념에 대해 학습하면 더 명확할 거 같네요. )

기준만 있으면 될 거 같습니다! 🙂

iampingu99 replied:
말씀해주신대로 목적에 따라서 트랜잭션 어노테이션을 잘 사용해야 할 것 같습니다. 👍

youngsu5582 commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.data.repository.CrudRepository;
+
+public interface ReservationRepository extends CrudRepository<Reservation, Long> {
+    @Query("SELECT r FROM Reservation r JOIN FETCH r.theme JOIN FETCH r.time")
```
EntityGraph 키워드를 사용하면, 더 간편하게 될 거 같네용👍

iampingu99 replied:
위의 코드에서 EntityGraph 키워드를 사용하니 더 직관적이네요 추가로 학습해서 필요한 곳에 잘 적용해보겠습니다 👍

youngsu5582 reviewed (changes requested):
이번에 리뷰 반영하면 Approve 하면 될 거 같네용 👍
좋은 질문해주셔서 저도 다시한번 저의 생각을 정리했네요. 감사합니당

youngsu5582 reviewed (approved):
이제, 4단계에서 원하는 요구사항은 다 된거 같아서 Approve 하겠습니다!
5~6단계 해서 올려주세용
